lib/sla.js

```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
  * @param {Object} inbound- the inbound channel
  * @param {Object} bridge - Object that defines the bridge to be passed to
  *   addChannelsToBridge().
+ * @ param {Object} trunkData - the data of the trunk to be used for station
+ *   access.
  * @return {Q} Q promise object.
  */
-function originateChannel(client, inbound, bridge) {
+function originateChannel(client, inbound, bridge, trunkData) {
   var defer = Q.defer();
-  var dialed = client.Channel();
-  var originate = Q.denodeify(dialed.originate.bind(dialed));
-  originate({endpoint: 'SIP/phone', app: 'sla', appArgs: 'dialed'})
-    .catch(function (err) {
-      err.name = 'originateFailure';
-      defer.reject(err);
-      
-    });
-  dialed.once('StasisStart', function(event, dialed) {
-    var answer = Q.denodeify(dialed.answer.bind(dialed));
+  var originate = [];
+  var numToHangup;
+  var dialed = [];
+
+  function dialedEnteredStasis(event, line) {
+    var answer = Q.denodeify(line.answer.bind(line));
     answer();
-    defer.resolve(dialed);
-  });
-  dialed.once('ChannelDestroyed', function(event, dialed) {
-    defer.resolve('Dialed channel hungup');
-  });
-  inbound.once('ChannelHangupRequest', function(event, inbound) {
-    var hangup = Q.denodeify(dialed.hangup.bind(dialed));
-    hangup();
+    var toKill = dialed.filter(function (unanswered) {
+      return unanswered.id !== line.id;
+    });
+    toKill.forEach(function (unanswered) {
+      unanswered.removeListener('ChannelDestroyed', dialedChannelHungup);
+      var hangup = Q.denodeify(unanswered.hangup.bind(unanswered));
+      hangup();
+    });
+    defer.resolve(line);
+  }
+
+  function dialedChannelHungup(event) {
+    numToHangup -= 1;
+    if(numToHangup === 0) {
+      inbound.removeListener('ChannelHangupRequest', inboundHungup);
+      defer.resolve('Dialed channels hungup');
+    }
+  }
+
+  function inboundHungup(event, line) {
+    dialed.forEach(function (unanswered) {
+      unanswered.removeListener('ChannelDestroyed', dialedChannelHungup);
+      var hangup = Q.denodeify(unanswered.hangup.bind(unanswered));
+      hangup();
+    });
     defer.resolve('Inbound channel hungup');
-  });
+  }
+
+  trunkData.getAllStations()
+    .then(function (stations) {
+      if (stations.length) {
+        stations.forEach(function() {
+          dialed.push(client.Channel());
```

Comment:
If you create an object with a station and a dialed channel, you'll be able to just iterate over that and not use an index anymore.